### PR TITLE
[Python UDF] Filter `NULL` values before calling the user defined function

### LINKED ISSUE #13138: python UDF: NULLs intercepted for scalars but not columnar args
### What happens?

According to https://duckdb.org/docs/api/python/function.html#null-handling, if you pass a NULL into a python UDF, duckdb will short-circuit and immediately resolve to NULL, and won't ever call the python with None.

This short-circuiting appears to happen when the input is a scalar, but doesn't when it is a column.

PS I think this short-circuiting (with a configurable override) is fantastic UX

### To Reproduce

```python
import duckdb
from duckdb.typing import INTEGER


def add_one(x):
    return x + 1


duckdb.create_function("add_one", add_one, [INTEGER], INTEGER, type="native")
duckdb.sql("SELECT add_one(1), add_one(NULL)").fetchall()
# [(2, None)]
duckdb.sql("SELECT add_one(x) FROM (SELECT UNNEST([1, NULL]) as x)").fetchall()
# Invalid Input Error: Python exception occurred while executing the UDF: TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'
```

### OS:

MacM1, MacOS 14.5

### DuckDB Version:

1.0.0

### DuckDB Client:

python

### Full Name:

Nick Crews

### Affiliation:

Ship Creek Group

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #13138 

This only applies to UDFs that use the default null handling (default behavior).

For both `native` and `arrow` UDF types, nulls are filtered out of the input that is sent to the user defined function.
On return to DuckDB the created result is converted to a DataChunk and the rows that were filtered out are replaced with NULL in the final result.

DuckDB side input:
```
┌─────────┐
│    a    │
│ varchar │
├─────────┤
│ 12      │
│ NULL    │
│ 42      │
└─────────┘
```

Received data by the UDF:
```
┌─────────┐
│    a    │
│ varchar │
├─────────┤
│ 12      │
│ 42      │
└─────────┘
```

In this example:
The resulting row produced by the UDF at (0 indexed) row 1 will be placed at row 2 in the final result.

### CODE REVIEW

File: tools/pythonpkg/tests/fast/udf/test_null_filtering.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        ],
+    )
+    def test_null_filtering_native(self, duckdb_cursor, table_data):
+        null_count = sum([1 for x in table_data if not x])
+        row_count = len(table_data)
+
+        df = pd.DataFrame({'a': table_data})
+        duckdb_cursor.execute("create table tbl as select * FROM df")
+
+        def my_native_func(x):
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/udf/test_null_filtering.py:30]: Can we add some tests with functions that receive multiple input parameters as well?

File: tools/pythonpkg/src/python_udf.cpp [hunk 2]
```diff
@@ -154,35 +187,74 @@ static scalar_function_t CreateVectorizedFunction(PyObject *function, PythonExce
 			}
 		}
 		// Convert the pyarrow result back to a DuckDB datachunk
-		ConvertPyArrowToDataChunk(python_object, result, state.GetContext(), count);
+		if (count != input_size) {
+			// We filtered out some NULLs, now we need to reconstruct the final result by adding the nulls back
+			Vector temp(result.GetType(), count);
+			// Convert the table into a temporary Vector
+			ConvertPyArrowToDataChunk(python_object, temp, state.GetContext(), count);
+			SelectionVector inverted(input_size);
```
Comment (Mytherin) [inline tools/pythonpkg/src/python_udf.cpp:195]: `ConvertPyArrowToDataChunk` should also not return any NULL values here - perhaps we can add a test & add verification that no NULL values are returned? Then we can throw an error message instructing the user to change the `NULL_HANDLING` in that case.

File: tools/pythonpkg/src/python_udf.cpp [hunk 3]
```diff
@@ -154,35 +187,74 @@ static scalar_function_t CreateVectorizedFunction(PyObject *function, PythonExce
...
 			}
+			if (contains_null) {
+				// Immediately insert None, no need to call the function
+				python_objects.push_back(py::none());
+				python_results[row] = py::none().ptr();
+				continue;
+			}
 
 			// Call the function
 			auto ret = PyObject_CallObject(function, bundled_parameters.ptr());
```
Comment (Mytherin) [inline tools/pythonpkg/src/python_udf.cpp:260]: Same here - `ret` should not be `Py_None` here

File: tools/pythonpkg/src/python_udf.cpp (hunk 2 above)
Comment (Tishj) [reply tools/pythonpkg/src/python_udf.cpp:195]: I believe that's done upstream by `VerifyNullHandling`, but that is DEBUG code, I assume you want this done on release builds?

Comment (Mytherin) [reply tools/pythonpkg/src/python_udf.cpp:195]: Yes I think doing this on release builds makes sense since this is user input, not an internal function created by us.

**Comment by Mytherin** [comment]
Thanks!
